pkg/monitor/ovsdb.go

```diff
@@ -160,411 +162,424 @@ func (monitor *OVSDBMonitor) startOvsdbMonitor() error {
-	ofPort, ok := localEndpointIface.Fields["ofport"].(float64)
+	oldEndpoint, ok := monitor.endpointMap[uuid]
 	if !ok {
-		return nil, nil
-	}
-	if ofPort <= 0 {
-		monitor.localEndpointHardwareAddrCacheLock.Lock()
-		ofport = monitor.localEndpointHardwareAddrCache[externalIDs[LocalEndpointIdentity].(string)]
-		monitor.localEndpointHardwareAddrCacheLock.Unlock()
-	} else {
-		ofport = uint32(ofPort)
+		return
 	}
 
-	newEndpoint := monitor.interfaceToEndpoint(ofport, portName, externalIDs[LocalEndpointIdentity].(string))
-	oldEndpoint := &datapath.Endpoint{
-		InterfaceName: newEndpoint.InterfaceName,
-		MacAddrStr:    newEndpoint.MacAddrStr,
-		PortNo:        newEndpoint.PortNo,
-		BridgeName:    newEndpoint.BridgeName,
-		VlanID:        uint16(oldTag),
+	if monitor.isEndpointReady(oldEndpoint) {
+		monitor.ovsdbEventHandler.DeleteLocalEndpoint(monitor.endpointMap[uuid])
 	}
-
-	return newEndpoint, oldEndpoint
+	delete(monitor.endpointMap, uuid)
 }
 
-func (monitor *OVSDBMonitor) isLocalEndpointPort(portName string) (bool, ovsdb.Row) {
-	for i, iface := range monitor.ovsdbCache["Interface"] {
-		if iface.Fields["name"].(string) == portName {
-			if _, ok := iface.Fields["external_ids"].(ovsdb.OvsMap).GoMap[LocalEndpointIdentity]; ok {
-				return true, monitor.ovsdbCache["Interface"][i]
+func (monitor *OVSDBMonitor) getPortBridgeName(portUUID string) string {
+	var bridgeName string
+	for _, bridge := range monitor.ovsdbCache["Bridge"] {
+		portUUIDs := listUUID(bridge.Fields["ports"])
+		for _, uuid := range portUUIDs {
+			if uuid.GoUuid == portUUID {
+				bridgeName = bridge.Fields["name"].(string)
+				return bridgeName
 			}
 		}
 	}
-	return false, ovsdb.Row{}
-}
-
-func (monitor *OVSDBMonitor) processPortUpdate(rowupdate ovsdb.RowUpdate) {
-	newEndpoint, oldEndpoint := monitor.filterPortVlanModeUpdate(rowupdate)
-	if newEndpoint != nil && oldEndpoint != nil {
-		go monitor.ovsdbEventHandler.UpdateLocalEndpoint(newEndpoint, oldEndpoint)
-	} else {
-		newEndpoint, oldEndpoint = monitor.filterPortVlanTagUpdate(rowupdate)
-		if newEndpoint != nil && oldEndpoint != nil {
-			go monitor.ovsdbEventHandler.UpdateLocalEndpoint(newEndpoint, oldEndpoint)
-		}
-	}
-}
 
-func (monitor *OVSDBMonitor) processEndpointUpdate(rowupdate ovsdb.RowUpdate) {
-	newEndpoint, oldEndpoint := monitor.filterEndpoint(rowupdate)
-	if newEndpoint != nil && oldEndpoint != nil {
-		go monitor.ovsdbEventHandler.UpdateLocalEndpoint(newEndpoint, oldEndpoint)
-	} else if newEndpoint != nil {
-		go monitor.ovsdbEventHandler.AddLocalEndpoint(newEndpoint)
-	}
+	return bridgeName
 }
 
-func (monitor *OVSDBMonitor) processEndpointDel(rowupdate ovsdb.RowUpdate) {
-	deletedEndpoints := monitor.filterEndpointDeleted(rowupdate)
-	if deletedEndpoints != nil {
-		go monitor.ovsdbEventHandler.DeleteLocalEndpoint(deletedEndpoints)
-	}
+func (monitor *OVSDBMonitor) isEndpointReady(endpoint *datapath.Endpoint) bool {
+	return endpoint.BridgeName != "" && endpoint.InterfaceUUID != "" &&
```

Comment:
Do you need  to check the endpoint is nil or not?

Reply:
We guarantee that the value passed in will not be nil at where the function is called.